### PR TITLE
CODE: rework threading on QueryStreamWriter

### PR DESCRIPTION
**Pull request by bluemonk3y**
#382 drop threading in QueryStreamWriter. @dguy  - can you LMK thoughts on the BlockingQueue

### CODE REVIEW

**Comment by bluemonk3y** [comment]
retest this please

File: ksql-engine/src/main/java/io/confluent/ksql/structured/QueuedSchemaKStream.java [hunk 1]
```diff
@@ -32,15 +32,13 @@
 import org.apache.kafka.streams.kstream.KStream;
 import org.apache.kafka.streams.kstream.Windowed;
 
-import java.util.List;
-import java.util.Objects;
-import java.util.Optional;
-import java.util.Set;
-import java.util.concurrent.SynchronousQueue;
+import java.util.*;
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/structured/QueuedSchemaKStream.java]: imports

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/QueryStreamWriter.java [hunk 2]
```diff
@@ -61,82 +65,62 @@
...
+    output.write("\n".getBytes());
+    output.flush();
+  }
+
   private class StreamsExceptionHandler implements Thread.UncaughtExceptionHandler {
+
     @Override
     public void uncaughtException(Thread thread, Throwable exception) {
-      streamsException.compareAndSet(null, exception);
+      streamsException = exception;
```
Comment (dguy) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/QueryStreamWriter.java]: if we get a StreamsException the only sensible thing to do really is shutdown

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/QueryStreamWriter.java [hunk 3]
```diff
@@ -61,82 +65,62 @@
...
-          log.warn(
-              "Failed to join row writer thread; setting to daemon to avoid hanging on shutdown"
-          );
-          rowWriterThread.setDaemon(true);
+        if (streamsException != null) {
+          throw streamsException;
         }
       }
+    } catch (EOFException exception) {
+      // The user has terminated the connection; we can stop writing
```
Comment (dguy) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/QueryStreamWriter.java]: Should we at least log this so we know why it has been terminated?

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/QueryStreamWriter.java [hunk 4]
```diff
@@ -61,82 +65,62 @@
...
+    } catch (InterruptedException exception) {
+      // The most likely cause of this is the server shutting down. Should just try to close
+      // gracefully, without writing any more to the connection stream.
+      log.warn("Interrupted while writing to connection stream");
+    } catch (Throwable exception) {
+      log.error("Exception occurred while writing to connection stream: ", exception);
+      out.write("\n".getBytes());
+      if (exception.getCause() instanceof KsqlException) {
+        new ObjectMapper().writeValue(out, new StreamedRow(exception.getCause()));
```
Comment (dguy) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/QueryStreamWriter.java]: any reason why we don't use the `objectMapper` field here and below?

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java [hunk 5]
```diff
@@ -97,4 +81,27 @@ public Response streamQuery(KsqlRequest request) throws Exception {
       ));
     }
   }
+
+  private TopicStreamWriter getTopicStreamWriter(Map<String, Object> clientLocalProperties, PrintTopic statement) {
+    PrintTopic printTopic = statement;
```
Comment (dguy) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java]: `final` params. Also, why assign the statement to the `printTopic`? looks redundant

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java [hunk 6]
```diff
@@ -97,4 +81,27 @@ public Response streamQuery(KsqlRequest request) throws Exception {
...
+    String topicName = printTopic.getTopic().toString();
```
Comment (dguy) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java]: Is this right? the topic name would be "topic something". Maybe add a `topicName` method to `PrintTopic`?

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java (hunk 5 above)
Comment (bluemonk3y) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java]: done

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/QueryStreamWriter.java (hunk 2 above)
Comment (bluemonk3y) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/QueryStreamWriter.java]: it gets passed back to the stream loop above,  that loop will rethrow and cleanup

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java (hunk 6 above)
Comment (bluemonk3y) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java]: printTopic.toString() would give you ' topic stuff' - getTopic returns QualifiedTopic - which implements  toString with correct formatting. I removed the redundant reference on the line below

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/QueryStreamWriter.java (hunk 4 above)
Comment (bluemonk3y) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/QueryStreamWriter.java]: done

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/QueryStreamWriter.java (hunk 3 above)
Comment (bluemonk3y) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/QueryStreamWriter.java]: done

File: ksql-engine/src/main/java/io/confluent/ksql/structured/QueuedSchemaKStream.java (hunk 1 above)
Comment (bluemonk3y) [reply ksql-engine/src/main/java/io/confluent/ksql/structured/QueuedSchemaKStream.java]: done

**Pull request merged**